### PR TITLE
WIP: API endpoint to List OSPs

### PR DESCRIPTION
**Pull request by ahmedwaleedmalik**
**What does this PR do / Why do we need it**:

This PR does the following:
- Grant OSM service account access to get,list OSPs from `kubermatic` namespace
- Introduce `/api/v2/seeds/{seed_name}/operatingsystemprofiles` to list OSPs

**Does this PR close any issues?**:
Fixes #

**Special notes for your reviewer**:

**Documentation**:

**Does this PR introduce a user-facing change?**:

```release-note
Add endpoint to list operating system profiles at the seed level.
```

### CODE REVIEW

File: pkg/controller/seed-controller-manager/kubernetes/resources.go [hunk 1]
```diff
@@ -533,10 +534,20 @@ func (r *Reconciler) ensureRoles(ctx context.Context, c *kubermaticv1.Cluster) e
...
+	if c.Spec.EnableOperatingSystemManager {
+		namedRoleCreatorGetters := []reconciling.NamedRoleCreatorGetter{
+			operatingsystemmanager.KubermaticOperatingSystemProfileRoleCreator,
+		}
+
+		if err := reconciling.ReconcileRoles(ctx, namedRoleCreatorGetters, seedNamespace, r.Client); err != nil {
+			return fmt.Errorf("failed to ensure RoleBindings: %w", err)
+		}
+	}
+
```
Comment (ahmedwaleedmalik) [inline pkg/controller/seed-controller-manager/kubernetes/resources.go]: @xrstf PTAL at this and let me know what you think

File: pkg/handler/v2/operatingsystemprofile/operatingsystemprofile.go [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+	{
+		Name: "osp-rockylinux",
+	},
+	{
+		Name: "osp-sles",
+	},
+	{
+		Name: "osp-ubuntu",
+	},
+}
```
Comment (ahmedwaleedmalik) [inline pkg/handler/v2/operatingsystemprofile/operatingsystemprofile.go:61]: I hate this part and I thought quite a lot before doing this. But tbh I couldn't think of any better solution for this.

File: pkg/provider/kubernetes/operating_system_profiles.go [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+	osmv1alpha1 "k8c.io/operating-system-manager/pkg/crd/osm/v1alpha1"
+
+	"k8s.io/apimachinery/pkg/api/meta"
+	ctrlruntimeclient "sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// PrivilegedOperatingSystemProfileProvider struct that holds required components of the PrivilegedOperatingSystemProfileProvider.
+type PrivilegedOperatingSystemProfileProvider struct {
+	privilegedClient ctrlruntimeclient.Client
+	namespace        string
```
Comment (lsviben) [inline pkg/provider/kubernetes/operating_system_profiles.go:33]: the namespace is not used

Comment (ahmedwaleedmalik) [reply pkg/provider/kubernetes/operating_system_profiles.go:33]: Good catch, let me fix that.

**Comment by lsviben** [comment]
I dont know OSM TBH, but for me it looks like you could avoid these changes for creating roles/rolebindings in seed namespace for every user cluster OSM if the OSP would be cluster-scoped. Is there some particular reason why its not?

**Comment by ahmedwaleedmalik** [comment]
> I dont know OSM TBH, but for me it looks like you could avoid these changes for creating roles/rolebindings in seed namespace for every user cluster OSM if the OSP would be cluster-scoped. Is there some particular reason why its not?

I understand your concern. OSM is deployed in the user cluster namespace, just like machine-controller. And it creates the CRs(not CRDs) for OperatingSystemProfiles on its own, in the user cluster namespace.

If someone wants to use those default OSPs, the ones I have hardcoded, everything works fine. They don't even need anything from the API tbh since the MC webhooks will take care of this defaulting.

But one of the core features of OSM is to allow users to consume custom OSPs. These OSPs should exist before cluster creation so we can't use the user cluster namespace for that. This is the reason why we chose the seed namespace aka `kubermatic` namespace and then added this RBAC.

**Comment by ahmedwaleedmalik** [comment]
/hold 

Some more changes are required for RBAC.

**Comment by lsviben** [comment]
> > I dont know OSM TBH, but for me it looks like you could avoid these changes for creating roles/rolebindings in seed namespace for every user cluster OSM if the OSP would be cluster-scoped. Is there some particular reason why its not?
> 
> I understand your concern. OSM is deployed in the user cluster namespace, just like machine-controller. And it creates the CRs(not CRDs) for OperatingSystemProfiles on its own, in the user cluster namespace.
> 
> If someone wants to use those default OSPs, the ones I have hardcoded, everything works fine. They don't even need anything from the API tbh since the MC webhooks will take care of this defaulting.
> 
> But one of the core features of OSM is to allow users to consume custom OSPs. These OSPs should exist before cluster creation so we can't use the user cluster namespace for that. This is the reason why we chose the seed namespace aka `kubermatic` namespace and then added this RBAC.
> 
> Cluster scoped OSPs are not possible since users can override and use custom version of OSM, just like MC, for user clusters as well.

Ok I see, didnt know that there are also OSPs in the user-cluster-namespace. These in the Seed namespace(kubermatic) are "custom" ones. Understood

File: pkg/controller/seed-controller-manager/kubernetes/resources.go [hunk 4]
```diff
@@ -533,10 +534,20 @@ func (r *Reconciler) ensureRoles(ctx context.Context, c *kubermaticv1.Cluster) e
 		return fmt.Errorf("failed to ensure Roles: %w", err)
 	}
 
+	if c.Spec.EnableOperatingSystemManager {
+		namedRoleCreatorGetters := []reconciling.NamedRoleCreatorGetter{
+			operatingsystemmanager.KubermaticOperatingSystemProfileRoleCreator,
+		}
+
+		if err := reconciling.ReconcileRoles(ctx, namedRoleCreatorGetters, seedNamespace, r.Client); err != nil {
+			return fmt.Errorf("failed to ensure RoleBindings: %w", err)
```
Comment (xrstf) [inline pkg/controller/seed-controller-manager/kubernetes/resources.go:543]: s/Binding//

File: pkg/resources/operatingsystemmanager/rbac.go [hunk 5]
```diff
@@ -24,9 +24,11 @@ import (
 )
 
 const (
-	serviceAccountName = "kubermatic-operating-system-manager"
-	roleName           = "kubermatic:operating-system-manager"
-	roleBindingName    = "kubermatic:operating-system-manager"
+	serviceAccountName                              = "kubermatic-operating-system-manager"
+	roleName                                        = "kubermatic:operating-system-manager"
+	roleBindingName                                 = "kubermatic:operating-system-manager"
+	kubermaticOperatingSystemProfileroleName        = "kubermatic:operating-system-profile-viewer"
```
Comment (xrstf) [inline pkg/resources/operatingsystemmanager/rbac.go:32]: That const name is a bit redundant, maybe `ospViewerRoleName` would be enough?

**Comment by ahmedwaleedmalik** [comment]
Closing this PR in favor of https://github.com/kubermatic/kubermatic/pull/10532

**Pull request closed without merging**